File: 2347-count-nodes-equal-to-average-of-subtree/count-nodes-equal-to-average-of-subtree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution
{
    int count=0;
    int[] dfs(TreeNode root)
    {
        if(root==null)
            return new int[]{0,0};
        
        int left[] = dfs(root.left);
        int right[] = dfs(root.right);

        int sum = root.val + left[0] + right[0];
        int n = 1 + left[1] + right[1];

        if(sum/n == root.val)
            count++;
        
        return new int[] {sum,n};
    }
    public int averageOfSubtree(TreeNode root)
    {
        dfs(root);
        return count;
    }
}

File: 2347-count-nodes-equal-to-average-of-subtree/count-nodes-equal-to-average-of-subtree.cpp
class Solution
{
public:

    int count = 0;

    pair<int,int> dfs(TreeNode* root)
    {
        if(root == NULL)
            return {0, 0};

        pair<int,int> left = dfs(root->left);
        pair<int,int> right = dfs(root->right);

        int sum = root->val + left.first + right.first;
        int nodes = 1 + left.second + right.second;

        if(sum / nodes == root->val)
            count++;

        return {sum, nodes};
    }

    int averageOfSubtree(TreeNode* root)
    {
        dfs(root);
        return count;
    }
};